stl/inc/span

```diff
@@ -0,0 +1,778 @@
+                                 && !_Is_std_array_v<remove_cvref_t<_Range>>
+                                 && _RANGES contiguous_range<_Range>
+                                 && _RANGES sized_range<_Range>
+                                 && (_RANGES _Forwarding_range<_Range> || is_const_v<_Ty>)
+                                 && is_convertible_v<remove_reference_t<_RANGES range_reference_t<_Range>>(*)[], _Ty(*)[]>;
+// clang-format on
+#else // ^^^ __cpp_lib_concepts / !__cpp_lib_concepts vvv
+
+// STRUCT TEMPLATE _Is_span
+template <class>
+struct _Is_span : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_span<span<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_std_array
+template <class>
+struct _Is_std_array : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_std_array<array<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Range, class _Ty>
+struct _Is_span_convertible_range
+    : bool_constant<is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Range&>()))> (*)[], _Ty (*)[]>> {
+};
+
+// STRUCT TEMPLATE _Has_container_interface
+template <class, class = void>
+struct _Has_container_interface : false_type {};
+
+template <class _Range>
+struct _Has_container_interface<_Range,
+    void_t<decltype(_STD data(_STD declval<_Range&>())), decltype(_STD size(_STD declval<_Range&>()))>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_compatible_range
+// clang-format off
+template <class _Range, class _Ty>
+struct _Is_span_compatible_range : bool_constant<conjunction_v<
+    negation<is_array<_Range>>,
+    negation<_Is_span<remove_const_t<_Range>>>,
+    negation<_Is_std_array<remove_const_t<_Range>>>,
+    _Has_container_interface<_Range>,
+    _Is_span_convertible_range<_Range, _Ty>>> {};
+// clang-format on
+
+#endif // !__cpp_lib_concepts
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using size_type              = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr size_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+#ifdef __cpp_lib_concepts
+
+    // clang-format off
+    template <size_t _Ext = _Extent>
+        requires(_Ext == 0 || _Ext == dynamic_extent)
+    constexpr span() noexcept {}

```

Comment:
A *requires-clause* can constrain a non-template, which means this can be a one-liner:
```c++
constexpr span() noexcept requires(_Extent == 0 || _Extent == dynamic_extent) = default;
```

Reply:
Thanks a lot. That is much better